File: src/main/java/isop/domain/TypeContainer.java
package isop.domain;

import java.util.HashMap;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author mathieu
 */
public class TypeContainer implements Function<Class,Object> {
    public TypeContainer()
    {
        instances= new HashMap<>();
    }
    
    private HashMap<Class, Object> instances;
    private Function<Class,Object> factory;
    public void add(Class t,Object instance){
        instances.put(t,instance);
    }

    public Object apply(Class c) {
        if (!instances.containsKey(c)) {
            if (factory != null) {
                instances.put(c, factory.apply(c));
            } else {
                try {
                    instances.put(c, c.newInstance());
                } catch (InstantiationException | IllegalAccessException ex) {
                    Logger.getLogger(TypeContainer.class.getName()).log(Level.SEVERE, null, ex);
                    throw new RuntimeException(ex);
                }
            }
        }
        return instances.get(c);
    }

    public void setFactory(Function<Class,Object> objectFactory) {
        this.factory = objectFactory;
    }
}

File: src/main/java/isop/domain/TypeContainer.kt
package isop.domain

import java.util.HashMap
import java.util.function.Function
import java.util.logging.Level
import java.util.logging.Logger

/**
 *
 * @author mathieu
 */
class TypeContainer : Function<Class<*>, Any?> {

    private val instances: HashMap<Class<*>, Any> = HashMap()
    private var factory: Function<Class<*>, Any>? = null

    fun add(t: Class<*>, instance: Any) {
        instances[t] = instance
    }

    override fun apply(c: Class<*>): Any? {
        if (!instances.containsKey(c)) {
            if (factory != null) {
                instances[c] = factory!!.apply(c)
            } else {
                try {
                    instances[c] = c.newInstance()
                } catch (ex: InstantiationException) {
                    Logger.getLogger(TypeContainer::class.java.name).log(Level.SEVERE, null, ex)
                    throw RuntimeException(ex)
                } catch (ex: IllegalAccessException) {
                    Logger.getLogger(TypeContainer::class.java.name).log(Level.SEVERE, null, ex)
                    throw RuntimeException(ex)
                }

            }
        }
        return instances[c]
    }

    fun setFactory(objectFactory: Function<Class<*>, Any>) {
        this.factory = objectFactory
    }
}
